File: android/src/main/java/expo/modules/rnnetwork/RnNetworkModule.kt
package expo.modules.rnnetwork

import com.scotia.rnnetwork.contracts.RNNetworkRegistry
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONArray
import org.json.JSONObject

class RNNetworkModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("RNNetworkModule")

        Events("sessionExpired")

        OnCreate {
            // Bridge the native session-expired hook to a JS event. The host invokes
            // RNNetworkRegistry.onSessionExpired?.invoke() when it detects the session is gone.
            RNNetworkRegistry.onSessionExpired = {
                this@RNNetworkModule.sendEvent("sessionExpired", emptyMap<String, Any?>())
            }
        }

        OnDestroy {
            RNNetworkRegistry.onSessionExpired = null
        }

        Function("hasNativeProvider") {
            RNNetworkRegistry.provider != null
        }

        Function("debugIdentity") {
            "registryId=${System.identityHashCode(RNNetworkRegistry)} " +
                "classloader=${RNNetworkRegistry::class.java.classLoader}"
        }

        Function("getNativeAppConfig") {
            RNNetworkRegistry.appConfig?.let { c ->
                mapOf(
                    "country" to c.country,
                    "environment" to c.environment,
                    "domains" to c.domains.map { mapOf("key" to it.key, "baseURL" to it.baseURL) },
                )
            }
        }

        Function("getNativeActiveDomain") {
            RNNetworkRegistry.activeDomain
        }

        Function("getBaseURLForDomain") { domainKey: String ->
            RNNetworkRegistry.appConfig?.domains?.firstOrNull { it.key == domainKey }?.baseURL
        }

        AsyncFunction("setActiveDomain") { domainKey: String ->
            val domains = RNNetworkRegistry.appConfig?.domains ?: return@AsyncFunction
            if (domains.none { it.key == domainKey }) return@AsyncFunction
            RNNetworkRegistry.activeDomain = domainKey
        }

        AsyncFunction("cancel") { requestId: String ->
            RNNetworkRegistry.provider?.cancel(requestId)
        }

        AsyncFunction("request") Coroutine { requestId: String, url: String, method: String, headers: Map<String, String>, body: Map<String, Any?>? ->
            val provider = RNNetworkRegistry.provider
                ?: throw NetworkException("PROVIDER_NOT_SET", retryable = false)

            val response = try {
                provider.request(requestId, url, method, headers, body)
            } catch (e: Throwable) {
                throw NetworkErrorMapper.map(e)
            }

            // Central rule: only 2xx is success. The host can't "forget" to throw anymore.
            if (response.statusCode !in 200..299) {
                val retryable = response.statusCode >= 500
                val code = if (response.statusCode < 500) "HTTP_CLIENT_ERROR" else "HTTP_SERVER_ERROR"
                throw NetworkException(code, retryable = retryable, httpStatus = response.statusCode)
            }

            val data = response.data
            val bodyMap: Map<String, Any?> = if (data == null || data.isEmpty()) {
                emptyMap()
            } else {
                try {
                    jsonToMap(JSONObject(String(data, Charsets.UTF_8)))
                } catch (e: Exception) {
                    throw NetworkException("INVALID_RESPONSE_BODY", retryable = false)
                }
            }

            mapOf(
                "body" to bodyMap,
                "statusCode" to response.statusCode,
                "headers" to response.headers,
            )
        }
    }
}

private fun jsonToMap(obj: JSONObject): Map<String, Any?> =
    obj.keys().asSequence().associateWith { key -> jsonValueToKotlin(obj.get(key)) }

private fun jsonToList(arr: JSONArray): List<Any?> =
    (0 until arr.length()).map { jsonValueToKotlin(arr.get(it)) }

private fun jsonValueToKotlin(value: Any): Any? = when (value) {
    JSONObject.NULL -> null
    is JSONObject   -> jsonToMap(value)
    is JSONArray    -> jsonToList(value)
    else            -> value
}

File: ios/RnNetworkModule.swift
import ExpoModulesCore
import iOSNetworkContract

public class RNNetworkModule: Module {
    public func definition() -> ModuleDefinition {
        Name("RNNetworkModule")

        Events("sessionExpired")

        OnCreate {
            // Bridge the native session-expired hook to a JS event. The host invokes
            // RNNetworkRegistry.onSessionExpired?() when it detects the session is gone
            // (e.g. after the app returns from background and the token is no longer valid).
            RNNetworkRegistry.onSessionExpired = { [weak self] in
                self?.sendEvent("sessionExpired", [:])
            }
        }

        OnDestroy {
            RNNetworkRegistry.onSessionExpired = nil
        }

        Function("hasNativeProvider") {
            return RNNetworkRegistry.provider != nil
        }

        Function("getNativeAppConfig") { () -> [String: Any]? in
            guard let c = RNNetworkRegistry.appConfig else { return nil }
            return [
                "country": c.country,
                "environment": c.environment,
                "domains": c.domains.map { ["key": $0.key, "baseURL": $0.baseURL] }
            ]
        }

        Function("getNativeActiveDomain") { () -> String? in
            RNNetworkRegistry.activeDomain
        }

        Function("getBaseURLForDomain") { (domainKey: String) -> String? in
            RNNetworkRegistry.appConfig?.domains.first { $0.key == domainKey }?.baseURL
        }

        AsyncFunction("setActiveDomain") { (domainKey: String) in
            guard let domains = RNNetworkRegistry.appConfig?.domains,
                  domains.contains(where: { $0.key == domainKey })
            else { return }
            RNNetworkRegistry.activeDomain = domainKey
        }

        AsyncFunction("cancel") { (requestId: String) in
            RNNetworkRegistry.provider?.cancel(requestId: requestId)
        }

        AsyncFunction("request") { (requestId: String, url: String, method: String, headers: [String: String], body: [String: Any]?) async throws -> [String: Any] in
            guard let provider = RNNetworkRegistry.provider else {
                throw NetworkException(code: "PROVIDER_NOT_SET", retryable: false)
            }

            let response: NetworkResponse
            do {
                response = try await provider.request(requestId: requestId, url: url, method: method, headers: headers, body: body)
            } catch {
                throw NetworkErrorMapper.map(error)
            }

            // Central rule: only 2xx is success. The host can't "forget" to throw anymore.
            guard (200...299).contains(response.statusCode) else {
                let retryable = response.statusCode >= 500
                let code = response.statusCode < 500 ? "HTTP_CLIENT_ERROR" : "HTTP_SERVER_ERROR"
                throw NetworkException(code: code, retryable: retryable, httpStatus: response.statusCode)
            }

            // Parse body (empty/204 → {}; non-JSON → error)
            let bodyDict: [String: Any]
            if let data = response.data, !data.isEmpty {
                guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                    throw NetworkException(code: "INVALID_RESPONSE_BODY", retryable: false)
                }
                bodyDict = json
            } else {
                bodyDict = [:]
            }

            return [
                "body": bodyDict,
                "statusCode": response.statusCode,
                "headers": response.headers,
            ]
        }
    }
}
